### PR TITLE
fix(gnovm): correct comparison between different types

### PR DESCRIPTION
**Pull request by ltzmaxwell**
Fixes https://github.com/gnolang/gno/issues/1376, fixes https://github.com/gnolang/gno/issues/2386.

### CODE REVIEW

**Comment by thehowl** [comment]
@ltzmaxwell can you merge master and fixup this PR for review? :)

**Comment by ltzmaxwell** [comment]
> @ltzmaxwell can you merge master and fixup this PR for review? :)

yes sure, in progress. @thehowl .

File: gnovm/pkg/gnolang/op_binary.go [hunk 1]
```diff
@@ -344,6 +343,9 @@ func isEql(store Store, lv, rv *TypedValue) bool {
 	} else if rvu {
 		return false
 	}
+	if err := checkSame(lv.T, rv.T, ""); err != nil {
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/op_binary.go:346]: there's an issue with this case:

```go
package main

func main() {
	var l interface{}
	l = int64(0)

	switch val := l.(type) {
	case int64, int:
		println(val)
		if val == 0 {
			println("l is zero")
		} else {
			println("NOT zero")
		}
	}
}
```


when we have a clause with multi cases, the type is val is set as the type of the switchStmt, in this case, `interface{}`,  when preprocess this `if val == 0 `, the type of the RHS is given it's default type `int`. this will lead to a `not same type` in runtime when checking `checkSame`,  and a incorrect result of the comparison.

one solution is to not convert the untyped const to its default type when it's an comparison expression and the other side is of type interface, and do the `convertUntypedTo` in runtime. but, is it correct?

File: examples/gno.land/p/demo/uassert/uassert_test.gno [hunk 2]
```diff
@@ -218,8 +218,8 @@ func TestEmpty(t *testing.T) {
 		{"", true},
 		{0, true},
 		{int(0), true},
-		{int64(0), true},
-		{uint(0), true},
+		{int64(0), false},
+		{uint(0), false},
 		// XXX: continue
```
Comment (ltzmaxwell) [inline examples/gno.land/p/demo/uassert/uassert_test.gno:223]: the original assertion is incorrect, see:
https://go.dev/play/p/B1PBlaB96Gd

**Comment by ltzmaxwell** [comment]
Hi @thehowl , this one is good for review. 🙏

File: gnovm/pkg/gnolang/op_expressions.go [hunk 3]
```diff
@@ -194,8 +194,13 @@ func (m *Machine) doOpRef() {
 			nv.Value = rv2
 		}
 	}
+	elt := xv.TV.T
+	if elt == DataByteType {
+		elt = xv.TV.V.(DataByteValue).ElemType
+	}
 	m.PushValue(TypedValue{
-		T: m.Alloc.NewType(&PointerType{Elt: xv.TV.T}),
+		// T: m.Alloc.NewType(&PointerType{Elt: xv.TV.T}),
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go:202]: Leftover?

File: examples/gno.land/p/demo/uassert/uassert_test.gno (hunk 2 above)
Comment (thehowl) [reply examples/gno.land/p/demo/uassert/uassert_test.gno:223]: Nice. Can you fix `uassert` to does what is expected? (return true even for int64(0)?)

A bit out of scope, but I think it's reasonable.

**Review by thehowl — APPROVED** [review]

File: gnovm/pkg/gnolang/op_expressions.go (hunk 3 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/op_expressions.go]: fixed.

File: examples/gno.land/p/demo/uassert/uassert_test.gno (hunk 2 above)
Comment (ltzmaxwell) [reply examples/gno.land/p/demo/uassert/uassert_test.gno]: the is fixed.

File: examples/gno.land/p/demo/uassert/uassert.gno [hunk 4]
```diff
@@ -380,10 +380,37 @@ func Empty(t TestingT, obj interface{}, msgs ...string) bool {
...
+	// XXX, this will result in a different result
+	// case uint64, int8:
+	// if val == 0 {}
+
+	// XXX: support other types like this:
+	// case int8:
+	// case int16:
+	// ...
+
 	case std.Address:
```
Comment (ltzmaxwell) [inline examples/gno.land/p/demo/uassert/uassert.gno:414]: fix uassert, add some comments.

File: examples/gno.land/p/demo/uassert/uassert.gno [hunk 5]
```diff
@@ -380,10 +380,37 @@ func Empty(t TestingT, obj interface{}, msgs ...string) bool {
 		if val != "" {
```
Comment (thehowl) [inline examples/gno.land/p/demo/uassert/uassert.gno:380]:
```go
func isNumberEmpty(n interface{}) (isNumber, isEmpty bool) {
	switch n := n.(type) {
	// NOTE: the cases are split individually, so that n becomes of the
	// asserted type; thus the comparison `n == 0` is concrete, and 0 is
	// not converted to interface{} and assumed to be an int.
	case int:
		return true, n == 0
	case int8:
		return true, n == 0
	case int16:
		return true, n == 0
	case int32:
		return true, n == 0
	case int64:
		return true, n == 0
	case uint:
		return true, n == 0
	case uint8:
		return true, n == 0
	case uint16:
		return true, n == 0
	case uint32:
		return true, n == 0
	case uint64:
		return true, n == 0
	case float32:
		return true, n == 0
	case float64:
		return true, n == 0
	}
	return false, false
}
```

then use it for both, before the type switch.

**Review by thehowl — APPROVED** [review]
LGTM.

cc @deelawn for second review.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 6]
```diff
@@ -194,8 +194,12 @@ func (m *Machine) doOpRef() {
 			nv.Value = rv2
 		}
 	}
+	elt := xv.TV.T
+	if elt == DataByteType {
```
Comment (deelawn) [inline gnovm/pkg/gnolang/op_expressions.go:198]: Can you put a comment here explaining why this is an exception?

File: gnovm/pkg/gnolang/preprocess.go [hunk 7]
```diff
@@ -1498,6 +1498,13 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				checkOrConvertIntegerKind(store, last, n.High)
 				checkOrConvertIntegerKind(store, last, n.Max)
 
+				// if n.X is untyped, convert to corresponding type
+				t := evalStaticTypeOf(store, last, n.X)
```
Comment (deelawn) [inline gnovm/pkg/gnolang/preprocess.go:1502]: Nice one 🔥  Untyped slices are not something I would have thought.

File: gnovm/tests/files/types/iface_eql.gno [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package main
+
+type Foo struct {
+	n int
+}
+
+func main() {
+	var l interface{}
+	//var l interface{} = 1
```
Comment (deelawn) [inline gnovm/tests/files/types/iface_eql.gno:9]: Are these supposed to be commented out?

File: gnovm/tests/files/types/cmp_iface_0_stdlibs.gno [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+type Error int64
+
+func (e Error) Error() string {
+	return "error: " + strconv.Itoa(int(e))
+}
+
+var errCmp = errors.New("XXXX")
+
+// specil case:
```
Comment (deelawn) [inline gnovm/tests/files/types/cmp_iface_0_stdlibs.gno:16]:
```suggestion
// special case:
```

File: gnovm/tests/files/types/cmp_iface_1.gno [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+type Error int64
+
+func (e Error) Error() string {
+	return "error: " + strconv.Itoa(int(e))
+}
+
+// typed
+var errCmp error = errors.New("XXXX")
+
+// specil case:
```
Comment (deelawn) [inline gnovm/tests/files/types/cmp_iface_1.gno:17]:
```suggestion
// special case:
```

**Review by deelawn — APPROVED** [review]
Nice job 👍 

I am surprised to see which test cases work every time I review one of these type PRs 😅 

I made a few suggestions for things to change, but looks good otherwise

File: gnovm/tests/files/types/iface_eql.gno (hunk 8 above)
Comment (ltzmaxwell) [reply gnovm/tests/files/types/iface_eql.gno:9]: fixed

File: gnovm/pkg/gnolang/op_expressions.go (hunk 6 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/op_expressions.go:199]: added.

File: examples/gno.land/p/demo/uassert/uassert.gno (hunk 5 above)
Comment (ltzmaxwell) [reply examples/gno.land/p/demo/uassert/uassert.gno]: updated, please double check. thank u.

File: gnovm/pkg/gnolang/preprocess.go [hunk 11]
```diff
@@ -2781,8 +2788,10 @@ func checkOrConvertType(store Store, last BlockNode, x *Expr, t Type, autoNative
 					// push t into bx.Left
 					checkOrConvertType(store, last, &bx.Left, t, autoNative)
 					return
-					// case EQL, LSS, GTR, NEQ, LEQ, GEQ:
-					// default:
+				case EQL, LSS, GTR, NEQ, LEQ, GEQ:
```
[the remark this answers is not in the document]
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:2791]: That makes sense. However, personally, I'd like to keep these cases, as it provides a clearer overview of which situations need to be addressed, or not.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 12]
```diff
@@ -194,8 +194,13 @@ func (m *Machine) doOpRef() {
 			nv.Value = rv2
 		}
 	}
+	// when obtaining a pointer of the databyte type, use the ElemType of databyte
+	elt := xv.TV.T
+	if elt == DataByteType {
+		elt = xv.TV.V.(DataByteValue).ElemType
```
Comment (piux2) [inline gnovm/pkg/gnolang/op_expressions.go:200]: Nice work!  Everything works as expected. 

This line of code,  solve another issue that is not directly related to the comparison issues in this PR 

```
	a := []byte("hello")
	a1 := &a[1]
	b1 := *a1 
        println(b1)

interface conversion: gnolang.Value is nil, not gnolang.DataByteValue
```
there are some other things we need to take care of to solve the issue completely. We will write a separate PR

The rest LGTM

**Review by piux2 — APPROVED** [review]
